### PR TITLE
BUG: Series[int].__setitem__(mask, td64_or_dt64) incorrect casting

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -4550,11 +4550,16 @@ def putmask(self, mask, value):
...
 
         values = self._values.copy()
-        if isinstance(converted, np.timedelta64) and self.dtype == object:
+        dtype, _ = infer_dtype_from(converted, pandas_dtype=True)
+        if dtype.kind in ["m", "M"]:
             # https://github.com/numpy/numpy/issues/12550
             #  timedelta64 will incorrectly cast to int
-            converted = [converted] * mask.sum()
-            values[mask] = converted
+            if not is_list_like(converted):
```
Comment (jreback) [inline pandas/core/indexes/base.py:4557]: this pattern is pretty common, maybe a helper function?

File: pandas/core/internals/blocks.py [hunk 2]
```diff
@@ -1317,10 +1318,15 @@ def where(self, other, cond, errors="raise", axis: int = 0) -> List[Block]:
                 blocks = block.where(orig_other, cond, errors=errors, axis=axis)
                 return self._maybe_downcast(blocks, "infer")
 
-            elif isinstance(other, np.timedelta64):
-                # expressions.where will cast np.timedelta64 to int
-                result = self.values.copy()
-                result[~cond] = [other] * (~cond).sum()
+            dtype, _ = infer_dtype_from(other, pandas_dtype=True)
```
Comment (jreback) [inline pandas/core/internals/blocks.py:1321]: e.g here as well

**Review by jreback — APPROVED** [review]
lgtm a possible followup

File: pandas/core/indexes/base.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py:4557]: yah one more PR coming up that uses this pattern yet again; i think its going to end up in array_algos.putmask,

**Pull request merged**